src/java/org/apache/cassandra/auth/IAuthenticator.java

```diff
@@ -54,6 +57,17 @@ public interface IAuthenticator
      */
     void setup();
 
+    /**
+     * Allows custom authenticators to return a {@link AuthenticateMessage} with a known
+     * authenticator implementation (e.g. "com.datastax.bdp.cassandra.auth.DseAuthenticator"
+     * which enables SASL scheme negotiation) based on {@link Connection} and {@link ClientState}
+     * information.
+     */
+    default AuthenticateMessage getAuthenticateMessage(Connection connection, ClientState clientState)
```

Comment:
Thank you for considering my use case!  One thing I have been thinking about is I think we probably only need one of these parameters, but I'm not sure which.

* I've noticed that in this context, we're always dealing with a `ServerConnection` object (there's quite a bit of casting in other classes to inspect `ServerConnection`) that `ServerConnection` itself has a `getClientState()`.  So that could be an argument for just passing only a `ServerConnection` instance.
* `ClientState` itself could have everything we might want to inspect, and if it doesn't we could add to it (e.g. we could later add a `hasClientCertificates()` or something to this to meet my use case.

Reply:
@tolbertam I have wondered about the same. I don't know the code well enough to know if it's safe to pass `ServerConnection` here. 

From code inspection is see that we can create a `ClientState` without an `InetSocketAddress`. In this scenario, I would assume that we can't cast `Connection` as `ServerConnection` and hence cannot fetch `ClientState` from it. It was because of this assumption I've proposed adding `Connection` in addition to `ClientState`. This assumption can be wrong and I'm not sure at this stage how to validate it.

If we go with a single parameter, my proposal would be to use the object that causes less harm / contains just enough information. I believe that would be `ClientState` since we can already fetch information about driver and version and potentially add others there.

wdyt?

Reply:
Agreed,  I think passing only `ClientState` would be sufficient.  If someone finds that sufficiently lacking, they could update `ClientState` to include what they need, without having to change the `IAuthenticator` interface :+1: